### DISCUSSION TITLE
Different concepts to define an asset

### DISCUSSION THREAD

**Discussion by dridk** [dagster-io/dagster#16191] [Q&A] [answered]
Hi, 

I'm having a little trouble distinguishing the use of these different class to define an asset. 

- Asset Key :  briefly the name of the asset
- AssetSource: define an asset
- @asset : define an asset using a decorator. Is the decorated function converted to the AssetSource?
- AssetIn/AssetOut: an asset but as an input. why not using something like : 

```python
@asset( ins = {"x": AssetKey("y")})
def myasset(x):
     pass
```

**Comment by sryza** [comment, marked as answer]
> AssetSource: define an asset

There are two different Python classes that can represent assets:
- `AssetsDefinition` - contains specs for a set of assets, along with the function that knows how to materialize them
- `SourceAsset` - contains a specification for a single asset, without a function for materializing it.

> @asset : define an asset using a decorator. Is the decorated function converted to the AssetSource?

The decorated function is wrapped in an `AssetsDefinition`, which contains a single asset.

> AssetIn/AssetOut: an asset but as an input. why not using something like :

When specifying an asset or dependency, there's often other metadata that people want to attach than just the asset key. That said, we could potentially also allow what you're suggesting.
